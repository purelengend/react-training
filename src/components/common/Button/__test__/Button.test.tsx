import { customRender, customRenderer, screen } from '@src/test/test-utils';
import userEvent from '@testing-library/user-event';

import { Button } from '..';

describe('Button test cases', () => {
  it('should render correctly with children and be visible by default', () => {
    const mockButton = customRenderer(<Button>Mock Children</Button>);

    expect(mockButton.toJSON()).toMatchSnapshot();
  });

  it('should accept custom class name', () => {
    const mockClassName = 'mock-class-name';
    const mockButton = customRenderer(
      <Button className={mockClassName}>Mock Children</Button>
    );

    expect(mockButton.toJSON()).toMatchSnapshot();
  });

  it('should accept different type', () => {
    const mockSubmitButton = customRenderer(
      <Button type="submit">Mock Children</Button>
    );

    const mockResetButton = customRenderer(
      <Button type="reset">Mock Children</Button>
    );

    const mockButton = customRenderer(
      <Button type="button">Mock Children</Button>
    );

    expect(mockSubmitButton.toJSON()).toMatchSnapshot();
    expect(mockResetButton.toJSON()).toMatchSnapshot();
    expect(mockButton.toJSON()).toMatchSnapshot();
  });

  it('should be hidden', () => {
    const mockButton = customRenderer(
      <Button isVisible={false}>Mock Children</Button>
    );

    expect(mockButton.getInstance()).toBeNull();
  });

  it('should accept the custom dataId attribute', () => {
    const mockDataId = '55';

    const mockButton = customRenderer(
      <Button dataId={`${mockDataId}`}>Mock Children</Button>
    );

    expect(mockButton.toJSON()).toMatchSnapshot();
  });

  it('should accept custom onClick function and invoke it when being clicked', async () => {
    customRender(<Button>Mock Children</Button>);

    const mockOnClick = vi.fn();

    const mockButton = screen.getByRole('button');

    mockButton.onclick = mockOnClick;

    await userEvent.click(mockButton);

    expect(mockOnClick).toBeCalled();
  });
});
