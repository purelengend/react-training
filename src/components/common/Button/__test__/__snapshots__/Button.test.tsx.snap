// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button test cases > should accept custom class name 1`] = `
<button
  className="mock-class-name"
  data-id="0"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Mock Children
</button>
`;

exports[`Button test cases > should accept different type 1`] = `
<button
  data-id="0"
  disabled={false}
  onClick={[Function]}
  type="submit"
>
  Mock Children
</button>
`;

exports[`Button test cases > should accept different type 2`] = `
<button
  data-id="0"
  disabled={false}
  onClick={[Function]}
  type="reset"
>
  Mock Children
</button>
`;

exports[`Button test cases > should accept different type 3`] = `
<button
  data-id="0"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Mock Children
</button>
`;

exports[`Button test cases > should accept the custom dataId attribute 1`] = `
<button
  data-id="55"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Mock Children
</button>
`;

exports[`Button test cases > should render correctly with children and be visible by default 1`] = `
<button
  data-id="0"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Mock Children
</button>
`;
